src/Authoring/WinRT.SourceGenerator/Analyzers/ComImportInterfaceAnalyzer.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#if ROSLYN_4_12_0_OR_GREATER
+
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Operations;
+using WinRT.SourceGenerator;
+
+#nullable enable
+
+namespace Generator;
+
+/// <summary>
+/// A diagnostic analyzer to warn for casts to <see cref="System.Runtime.InteropServices.ComImportAttribute"/> interfaces.
+/// </summary>
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class ComImportInterfaceAnalyzer : DiagnosticAnalyzer
+{
+    /// <inheritdoc/>
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = [WinRTRules.ComImportInterfaceCast];
+
+    /// <inheritdoc/>
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(static context =>
+        {
+            // This analyzer is only enabled when 'EnableAotAnalyzer' is set and when the AOT warning level is not set to 0 (ie. disabled)
+            if (GeneratorExecutionContextHelper.GetCsWinRTAotWarningLevel(context.Options.AnalyzerConfigOptionsProvider) == 0 ||
```

Comment:
Can we also scope this to auto mode for now to avoid introducing warnings in non CsWinRT scenarios / opt-in scenarios.

`IsCsWinRTAotOptimizerInAutoMode`

Reply:
Updated!